### PR TITLE
ISPN-15332 Require AUTH during HELLO if authentication is enabled

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-15332

* Updates `HELLO` to require authentication parameters if needed;
* Updates `HELLO` version returned;
* Updates `CLIENT` to return for `SETINFO` sub-command;
* Update the tests to pass when updating to Lettuce 6.3.0.

I ran everything with 6.3.0 to assert it would work after the update in #11491.

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java [hunk 1]
```diff
@@ -23,7 +24,11 @@ public AUTH() {
    public CompletionStage<RespRequestHandler> perform(Resp3AuthHandler handler,
                                                       ChannelHandlerContext ctx,
                                                       List<byte[]> arguments) {
-      CompletionStage<Boolean> successStage = handler.performAuth(ctx, arguments.get(0), arguments.get(1));
+      CompletionStage<Boolean> successStage = handler.performAuth(ctx, arguments.get(0), arguments.get(1))
+            .thenApply(s -> {
+               if (s) Consumers.OK_BICONSUMER.accept(null, handler.allocator());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java:29]: The allocator should only be used in the callback from stageToReturn or without a lambda.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java [hunk 2]
```diff
@@ -35,25 +49,33 @@ public CompletionStage<RespRequestHandler> perform(Resp3AuthHandler handler,
...
+         successStage = handler.performAuth(ctx, arguments.get(2), arguments.get(3));
+      } else if (!handler.isAuthorized() && handler.canUseCertAuth()) {
+         successStage = handler.performAuth(ctx);
       } else {
-         if (arguments.size() == 4) {
-            successStage = handler.performAuth(ctx, arguments.get(2), arguments.get(3));
-         } else if (!handler.isAuthorized() && handler.canUseCertAuth()) {
-            successStage = handler.performAuth(ctx);
+         if (handler.isAuthenticationRequired()) {
+            RespErrorUtil.customRawError("NOAUTH HELLO must be called with the client already authenticated, otherwise the HELLO <proto> AUTH <user> <pass> option can be used to authenticate the client and select the RESP protocol version at the same time", handler.allocator());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java:61]: Is this the error that REDIS returns? I thought they normally have a dash as the first character?

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java [hunk 3]
```diff
@@ -35,25 +49,33 @@ public CompletionStage<RespRequestHandler> perform(Resp3AuthHandler handler,
...
+            return s;
+         }), ctx, success -> AUTH.createAfterAuthentication(success, handler));
       }
 
       return handler.myStage();
    }
 
    private static void helloResponse(ByteBufPool alloc) {
-      String versionString = Version.getBrandVersion();
+      String versionString = Version.getMajorMinor();
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java:78]: Man I really wish these were documented better as to what format you would expect.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/CLIENT.java [hunk 4]
```diff
@@ -65,6 +65,7 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, Channel
                      return handler.myStage();
                }
             }
+            Consumers.OK_BICONSUMER.accept(null, handler.allocator());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/CLIENT.java:68]: Lets add a test for SETINFO as well.

File: server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java [hunk 5]
```diff
@@ -45,8 +48,10 @@ static class Request {
 
    @BeforeClass
    public void beforeClass() {
+      RespServer server = mock();
```
Comment (wburns) [inline server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java:51]: Is this related? I don't see any code using the time service now.

**Review by wburns — CHANGES_REQUESTED** [review]

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/CLIENT.java (hunk 4 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/CLIENT.java:68]: We need Lettuce 6.3.0 for that :/

File: server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java (hunk 5 above)
Comment (jabolina) [reply server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java:51]: I am unsure what is different, but running the tests in the terminal, testng started to run the access log test at the beginning. If the access log runs concurrently with the decoder, it has the trace enabled and causes an NPE on:

https://github.com/infinispan/infinispan/blob/9e94ababc8e23c8f270e7974bd2cc0141e804274/server/resp/src/main/java/org/infinispan/server/resp/RespHandler.java#L90

I could change the handler directly, though. Seems to make more sense.

Comment (jabolina) [reply server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java:51]: By unsure what is different I mean I don't know how testng decides what to run :(

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java (hunk 3 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java:78]: This is simply to work with Lettuce. Version 6.3.0 started parsing the version string returned and expects it to have only numbers: https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/RedisHandshake.java#L322-L343

Let me add a comment there specifying that the version should be only numbers.

Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java:78]: I understand it is for Lettuce, I was more commenting wishing that our Version methods were documented with an example of what the String could look like.

We definitely need a comment here saying we need numbers only though.

File: server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java (hunk 5 above)
Comment (wburns) [reply server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java:51]: Aha, fun test leak issue. I would say we should really isolate the access logger tests themselves to prevent this type of thing, but I am fine with it this way.

File: server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java [hunk 6]
```diff
@@ -158,13 +158,13 @@ private <E> CompletionStage<RespRequestHandler> stageToReturn(CompletionStage<E>
             return myStage;
          }, ctx.channel().eventLoop());
       }
-      return stage.handle((value, t) -> {
+      return stage.handleAsync((value, t) -> {
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:161]: Yeah, I was thinking we would eventually have to do this as well.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java [hunk 7]
```diff
@@ -25,7 +26,10 @@ public CompletionStage<RespRequestHandler> perform(Resp3AuthHandler handler,
                                                       List<byte[]> arguments) {
       CompletionStage<Boolean> successStage = handler.performAuth(ctx, arguments.get(0), arguments.get(1));
 
-      return handler.stageToReturn(successStage, ctx, success -> createAfterAuthentication(success, handler));
+      return handler.stageToReturn(successStage, ctx, success -> {
+         if (success) Consumers.OK_BICONSUMER.accept(null, handler.allocator());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java:30]: I am not a fan of having this here and then the non success writing to the channel in handleAuthResponse. Is there a reason this was moved? We should strive to have the error and success paths when writing a response near each other.

File: server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java [hunk 8]
```diff
@@ -98,4 +97,8 @@ public boolean canUseCertAuth() {
       RespAuthenticator authenticator = respServer.getConfiguration().authentication().authenticator();
       return authenticator != null && authenticator.isClientCertAuthEnabled();
    }
+
+   public boolean isAuthenticationRequired() {
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:101]: Is this method not the inverse of `isAuthorized`?

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java (hunk 7 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/AUTH.java:30]: The mix with AUTH during the HELLO. In that case, we don't want the `+OK`, but we want the unauthorized response. Let me move some stuff around.

**Comment by karesti** [comment]
@jabolina why not updating to 6.3 here too ?

**Comment by jabolina** [comment]
@karesti I'll pull your commit in and add some tests.

**Comment by karesti** [comment]
I'm going to close the PR

**Review by karesti — APPROVED** [review]

**Comment by karesti** [comment]
thanks master of the universe @jabolina
